### PR TITLE
POC of a training setup

### PR DESCRIPTION
**Pull request by simlmx**
* Clear definitions of all the moving pieces
* Multi-processing of feature extraction (using "torchdata")
* Model interface is prod-friendly (the trained model can be called in inference without modification)
* Evaluation script + notebook to look at the results

### CODE REVIEW

File: psp/ml/metrics.py [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+def mean_absolute_error(y_true: Y, y_pred: Y) -> np.ndarray:
+    return abs(y_true.powers - y_pred.powers)
+
+
+class MeanRelativeError:
+    def __init__(self, cap: float | None = None):
+        self._cap = cap
+
+    def __call__(self, y_true: Y, y_pred: Y) -> np.ndarray:
+        err = abs(y_true.powers - y_pred.powers) / y_true.powers
```
Comment (peterdudfield) [inline psp/ml/metrics.py:21]: becasue of small numbers, do you want to consdier abs(y_true.powers - y_pred.powers) / (0.5*y_true.powers + 0.5*y_pred.powers)

File: psp/utils/profiling.py [hunk 2]
```diff
@@ -0,0 +1,10 @@
+import contextlib
+import time
+
+
+@contextlib.contextmanager
+def profile(name: str):
+    t0 = time.time()
+    yield
+    t1 = time.time()
+    print(f'Executed "{name}" in {t1 - t0:.3f}s')
```
Comment (peterdudfield) [inline psp/utils/profiling.py:10]: could be changed to `logging.info`?

File: psp/scripts/eval_model.py [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+    parser.add_argument(
+        "-n",
+        "--exp-name",
+        help="name of the experiment (used as the name of the dictionary)",
+    )
+    parser.add_argument("-s", "--sample", type=float, help="Sample the test samples.")
+    return parser.parse_args()
+
+
+def eval_model(args: argparse.Namespace):
```
Comment (peterdudfield) [inline psp/scripts/eval_model.py:33]: might want od move this inside the package, then the script just does the high level things

File: psp/scripts/eval_model.py [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+from psp.ml.models.previous_day import PreviousDayPvSiteModel
+from psp.ml.training import make_data_loader
+
+METRICS: dict[str, Metric] = {
+    "mre_cap=1": MeanRelativeError(cap=1),
+    "mae": mean_absolute_error,
+}
+
+
+def parse_args():
```
Comment (peterdudfield) [inline psp/scripts/eval_model.py:21]: check out [click](https://pypi.org/project/click/) - i think its much better

**Review by peterdudfield — COMMENTED** [review]
just added a few comments, lloks great. 

Would be great if you could use `ocf_datapipes` for PV data. Perhaps we can sit together and do a little demo

File: psp/ml/types.py [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+# TODO Is this the type we want to use for dates?
+# Maybe pandas' Timestamp?
+Timestamp = datetime.datetime
+
+
+@dataclasses.dataclass
+class X:
+    pv_id: PvId
+    # Time at which are making the prediction.
+    ts: Timestamp
```
Comment (peterdudfield) [inline psp/ml/types.py:19]: ts_utc

File: psp/ml/metrics.py (hunk 1 above)
Comment (simlmx) [reply psp/ml/metrics.py:21]: Good idea. I find that metric harder to interpret though! The `cap` option fixes the same issue I think? In general I'm not sure how much relative error is going to be useful anyway. MAE might be the way to go!
